### ISSUE TITLE
Won't work as expected: 920450

### ISSUE THREAD

**Issue by csanders-git** [SpiderLabs/owasp-modsecurity-crs#424]
As @dune73 has been so apt to teach us, in a chained rule not all elements in a collection that match will be sent to the next rule. In fact the entire first rule will run and if there is a match anywhere in it, then the chained rule will run. 

In rule 920450 we see this done perfectly incorrectly. As the following debug log will show:

```
[17/Jul/2016:18:59:38 --0400][/][4] Executing operator "rx" with param "^(.*)$" against REQUEST_HEADERS_NAMES:Host.
[17/Jul/2016:18:59:38 --0400][/][9] Added regex subexpression to TX.0: Host
[17/Jul/2016:18:59:38 --0400] [/][9] Setting variable: tx.header_name=/%{tx.0}/
...
[17/Jul/2016:18:59:38 --0400][/][4] Executing operator "rx" with param "^(.*)$" against REQUEST_HEADERS_NAMES:Content-Length.
[17/Jul/2016:18:59:38 --0400][/][9] Target value: "Content-Length"
[17/Jul/2016:18:59:38 --0400][/][9] Added regex subexpression to TX.0: Content-Length
[17/Jul/2016:18:59:38 --0400][/][9] Setting variable: tx.header_name=/%{tx.0}/
...
[17/Jul/2016:18:59:38 --0400][/][4] Executing operator "rx" with param "^(.*)$" against REQUEST_HEADERS_NAMES:User-Agent.
[17/Jul/2016:18:59:38 --0400][/][9] Target value: "User-Agent"
[17/Jul/2016:18:59:38 --0400][/][9] Added regex subexpression to TX.0: User-Agent
[17/Jul/2016:18:59:38 --0400][/][9] Setting variable: tx.header_name=/%{tx.0}/
...
[17/Jul/2016:18:59:38 --0400][/][4] Executing operator "rx" with param "^(.*)$" against REQUEST_HEADERS_NAMES:translated.
[17/Jul/2016:18:59:38 --0400][/][9] Target value: "translated"
[17/Jul/2016:18:59:38 --0400][/][9] Added regex subexpression to TX.0: translated
[17/Jul/2016:18:59:38 --0400][/][9] Setting variable: tx.header_name=/%{tx.0}/
...
[17/Jul/2016:18:59:38 --0400][/][4] Rule returned 1.
[17/Jul/2016:18:59:38 --0400][/][9] Match -> mode NEXT_RULE.
[17/Jul/2016:18:59:38 --0400][/][4] Recipe: Invoking rule 564e048500d0; [file "/etc/httpd/modsecurity.d/owasp-modsecurity-crs/rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf"] [line "1127"].
[17/Jul/2016:18:59:38 --0400] [/][5] Rule 564e048500d0: SecRule "TX:HEADER_NAME" "@within %{tx.restricted_headers}" "t:lowercase,setvar:tx.msg=%{rule.msg},setvar:tx.anomaly_score=+%{tx.critical_anomaly_score},setvar:tx.%{rule.id}-OWASP_CRS/POLICY/HEADERS_RESTRICTED-%{matched_var_name}=%{matched_var}"
[17/Jul/2016:18:59:38 --0400][/][9] T (0) lowercase: "/translated/"
[17/Jul/2016:18:59:38 --0400][/][4] Transformation completed in 14 usec.
[17/Jul/2016:18:59:38 --0400][/][4] Executing operator "within" with param "%{tx.restricted_headers}" against TX:header_name.
[17/Jul/2016:18:59:38 --0400][/][9] Target value: "/translated/"
[17/Jul/2016:18:59:38 --0400][/][9] Resolved macro %{tx.restricted_headers} to: content-range if lock-token translate
[17/Jul/2016:18:59:38 --0400][/][4] Operator completed in 21 usec.
[17/Jul/2016:18:59:38 --0400][/][4] Rule returned 0.
[17/Jul/2016:18:59:38 --0400][/][9] No match, not chained -> mode NEXT_RULE.
```

Now there are two options to get around this. One using a limited regex subset (see the DOS PR @dune73 submitted) and one with FPs using @within.

In this case I think the @within makes more sense. So the suggestion is as follows:

```
# 
# Restricted HTTP headers 
#
SecRule REQUEST_HEADERS_NAMES "@within %{tx.restricted_headers}" \
  "msg:'HTTP header is restricted by policy %{tx.0}',\
   severity:'CRITICAL',\
   phase:request,\
   t:none,\
   block,\
   rev:'2',\
   ver:'OWASP_CRS/3.0.0',\
   maturity:'9',\
   accuracy:'9',\
   id:920450,\
   logdata:'%{matched_var}',\
   tag:'application-multi',\
   tag:'language-multi',\
   tag:'platform-multi',\
   tag:'attack-protocol',\
   tag:'OWASP_CRS/POLICY/HEADER_RESTRICTED',\
   tag:'WASCTC/WASC-21',\
   tag:'OWASP_TOP_10/A7',\
   tag:'PCI/12.1',\
   tag:'WASCTC/WASC-15',\
   tag:'OWASP_TOP_10/A7',\
   tag:'PCI/12.1',\
   t:lowercase,\
   setvar:'tx.msg=%{rule.msg}',\
   setvar:tx.anomaly_score=+%{tx.critical_anomaly_score},\
   setvar:tx.%{rule.id}-OWASP_CRS/POLICY/HEADERS_RESTRICTED-%{matched_var_name}=%{matched_var}"
```

and to change setup.conf to have:

```
  setvar:'tx.restricted_headers=content-range if lock-token translate'"
```

 Now this will cause some False Positives for subsets of these headers... so a header 'trans' for instance would be blocked by this rule, but its pretty minor in this case.

**Comment by dune73** [comment]
Good catch.

Is not this something we want to fix before V3.0.0? I think it's a bug that should fixed after RC1, but before the final release.

I like the `@within` approach. I do not think the solution in PR #403 would work here, as the problem here is checking m request headers against n locally restricted headers. #403 it is 1 extension against n locally defined extensions.

There should not be too many FPs, but still, it does not feel very good. It's just that I do not see any other solution.

**Comment by csanders-git** [comment]
I put it as beyond 3 just so we can keep our final feature list. I'd like to fix it before 3.0 final

**Comment by dune73** [comment]
Why don't we do a V3.0.0-rc2 tag or milestone instead of using "Beyond v3.0.0" for something "Before v3.0.0"?

**Comment by csanders-git** [comment]
We can - I didn't want to put a tag on it at all.

**Comment by jschleus** [comment]
Hmm, I see many FPs with that change since now all partial content requests ("Range" request header) are  blocked since "range" matches "content-range" (maybe a new operator like "wordmatch" would be helpful).

**Comment by lifeforms** [comment]
@jschleus This should hopefully be fixed now in #491. Sorry for the problem.
